File: javascript/packages/engine/__tests__/engine/runtime/expressions/date/DefaultDateCalculator.test.ts
/*
 * Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { DefaultDateCalculator } from '../../../../../src/engine/runtime/expressions/date/DateCalculator'

function d(date: string): Date {
    return new Date(`${date}T00:00:00`)
}

describe('DefaultDateCalculator', () => {
    const calculator = new DefaultDateCalculator()
    const tz = Intl.DateTimeFormat().resolvedOptions().timeZone
    it('should create Date', () => {
        expect(calculator.createDate(2020, 1, 23)).toStrictEqual(new Date('2020-01-23T00:00:00'))
        expect(() => calculator.createDate(2020, 0, 1)).toThrow()

        expect(() => calculator.createDate(2020, 13, 1)).toThrow()
        expect(() => calculator.createDate(2020, 1, 0)).toThrow()

        expect(calculator.createDate('2020-01-23')).toStrictEqual(new Date('2020-01-23T00:00:00'))
        //expect(() => calculator.createDate("2020-02-30")).toThrow() // javascript allows values to overflow
        expect(() => calculator.createDate('2020-00-01')).toThrow()
        expect(() => calculator.createDate('2020-13-01')).toThrow()
        expect(() => calculator.createDate('2020-01-00')).toThrow()
        //expect(() => calculator.createDate("2020-01-23T10:00:00")).toThrow() // javascript allows values to overflow
        //expect(() => calculator.createDate("2020-01-23T10:00:00Z")).toThrow() // javascript allows values to overflow
    })
    it('should create DateTime', () => {
        expect(() => calculator.createDateTime('2020-01-23')).toThrow()
        expect(() => calculator.createDateTime('2020-01-23T10:00:00')).toThrow()
        // expect(() => calculator.createDateTime("2020-02-30T10:00:00Z")).toThrow() // javascript allows values to overflow
        expect(() => calculator.createDateTime('2020-00-01T10:00:00Z')).toThrow()
        expect(() => calculator.createDateTime('2020-13-01T10:00:00Z')).toThrow()
        expect(() => calculator.createDateTime('2020-01-00T10:00:00Z')).toThrow()
        expect(() => calculator.createDateTime('2020-01-23T25:00:00Z')).toThrow()
        expect(() => calculator.createDateTime('2020-01-23T10:60:00Z')).toThrow()
        expect(() => calculator.createDateTime('2020-01-23T10:00:60Z')).toThrow()

        expect(calculator.createDateTime('2020-01-23T10:00:00', tz)).toStrictEqual(new Date('2020-01-23T10:00:00'))
        expect(calculator.createDateTime('2020-01-23T10:00:00Z')).toStrictEqual(new Date('2020-01-23T10:00:00Z'))

        expect(calculator.createDateTime(2020, 1, 23, 10, 0, 0, tz)).toStrictEqual(new Date('2020-01-23T10:00:00'))
    })
    it('should return today', () => {
        const now = new Date()
        const today = new Date(now.getFullYear(), now.getMonth(), now.getDate())
        expect(calculator.today(tz)).toStrictEqual(today)
    })
    it('should convert Date to DateTime', () => {
        expect(calculator.toDateTime(d('2020-01-23'), tz)).toStrictEqual(new Date('2020-01-23T00:00:00'))
    })
    it('should convert DateTime to Date', () => {
        expect(calculator.toDate(d('2020-01-23'), tz)).toStrictEqual(new Date('2020-01-23T00:00:00'))
    })
    it('should get Date field', () => {
        const date = d('2020-01-30')
        expect(calculator.getDateField(date, 'YEAR')).toBe(2020)
        expect(calculator.getDateField(date, 'MONTH')).toBe(1)
        expect(calculator.getDateField(date, 'DAY_OF_MONTH')).toBe(30)
    })
    it('should set Date field', () => {
        const date = d('2020-01-30')
        expect(calculator.withDateField(date, 'YEAR', 2021)).toStrictEqual(new Date('2021-01-30T00:00:00'))
        expect(calculator.withDateField(date, 'MONTH', 3)).toStrictEqual(new Date('2020-03-30T00:00:00'))
        expect(calculator.withDateField(date, 'MONTH', 2)).toStrictEqual(new Date('2020-02-29T00:00:00'))
        expect(calculator.withDateField(date, 'DAY_OF_MONTH', 10)).toStrictEqual(new Date('2020-01-10T00:00:00'))

        expect(() => calculator.withDateField(date, 'MONTH', 13)).toThrow()
        expect(() => calculator.withDateField(date, 'DAY_OF_MONTH', 32)).toThrow()
    })
    it('should add Date field', () => {
        const date = d('2020-01-30')
        expect(calculator.addDateField(date, 'YEAR', 1)).toStrictEqual(new Date('2021-01-30T00:00:00'))
        expect(calculator.addDateField(date, 'MONTH', 1)).toStrictEqual(new Date('2020-02-29T00:00:00'))
        expect(calculator.addDateField(date, 'MONTH', 12)).toStrictEqual(new Date('2021-01-30T00:00:00'))
        expect(calculator.addDateField(date, 'DAY_OF_MONTH', 31)).toStrictEqual(new Date('2020-03-01T00:00:00'))
        expect(calculator.addDateField(date, 'DAY_OF_MONTH', 365)).toStrictEqual(new Date('2021-01-29T00:00:00'))
    })
    it('should diff Date field', () => {
        expect(calculator.differenceBetweenDates(d('2020-01-30'), d('2022-01-30'), 'YEAR')).toBe(2)
        expect(calculator.differenceBetweenDates(d('2020-01-30'), d('2022-01-29'), 'YEAR')).toBe(1)
        expect(calculator.differenceBetweenDates(d('2020-01-30'), d('2020-01-29'), 'YEAR')).toBe(0)

        expect(calculator.differenceBetweenDates(d('2020-01-29'), d('2020-02-29'), 'MONTH')).toBe(1)
        expect(calculator.differenceBetweenDates(d('2020-01-30'), d('2020-02-29'), 'MONTH')).toBe(0)
        expect(calculator.differenceBetweenDates(d('2020-01-30'), d('2020-03-01'), 'MONTH')).toBe(1)
        expect(calculator.differenceBetweenDates(d('2020-01-29'), d('2021-01-28'), 'MONTH')).toBe(11)

        expect(calculator.differenceBetweenDates(d('2020-01-29'), d('2020-02-29'), 'DAY_OF_MONTH')).toBe(31)
    })
    it('should get DateTime field', () => {
        const date = new Date('2020-01-30T10:10:13')
        expect(calculator.getDateTimeField(date, 'YEAR', tz)).toBe(2020)
        expect(calculator.getDateTimeField(date, 'MONTH', tz)).toBe(1)
        expect(calculator.getDateTimeField(date, 'DAY_OF_MONTH', tz)).toBe(30)
        expect(calculator.getDateTimeField(date, 'HOUR', tz)).toBe(10)
        expect(calculator.getDateTimeField(date, 'MINUTE', tz)).toBe(10)
        expect(calculator.getDateTimeField(date, 'SECOND', tz)).toBe(13)
    })
    it('should set DateTime field', () => {
        const date = new Date('2020-01-30T10:10:13')

        expect(calculator.withDateTimeField(date, 'YEAR', 2021, tz)).toStrictEqual(new Date('2021-01-30T10:10:13'))
        expect(calculator.withDateTimeField(date, 'MONTH', 3, tz)).toStrictEqual(new Date('2020-03-30T10:10:13'))
        expect(calculator.withDateTimeField(date, 'DAY_OF_MONTH', 10, tz)).toStrictEqual(
            new Date('2020-01-10T10:10:13'),
        )
        expect(calculator.withDateTimeField(date, 'HOUR', 23, tz)).toStrictEqual(new Date('2020-01-30T23:10:13'))
        expect(calculator.withDateTimeField(date, 'MINUTE', 5, tz)).toStrictEqual(new Date('2020-01-30T10:05:13'))
        expect(calculator.withDateTimeField(date, 'SECOND', 0, tz)).toStrictEqual(new Date('2020-01-30T10:10:00'))

        expect(calculator.withDateTimeField(date, 'MONTH', 2, tz)).toStrictEqual(new Date('2020-02-29T10:10:13'))

        expect(() => calculator.withDateTimeField(date, 'MONTH', 13, tz)).toThrow()
        expect(() => calculator.withDateTimeField(date, 'DAY_OF_MONTH', 32, tz)).toThrow()
        expect(() => calculator.withDateTimeField(date, 'HOUR', 24, tz)).toThrow()
        expect(() => calculator.withDateTimeField(date, 'MINUTE', 60, tz)).toThrow()
        expect(() => calculator.withDateTimeField(date, 'SECOND', 60, tz)).toThrow()
    })
    it('should add DateTime field', () => {
        const date = new Date('2020-01-30T10:10:13')

        expect(calculator.addDateTimeField(date, 'YEAR', 1, tz)).toStrictEqual(new Date('2021-01-30T10:10:13'))
        expect(calculator.addDateTimeField(date, 'MONTH', 1, tz)).toStrictEqual(new Date('2020-02-29T10:10:13'))
        expect(calculator.addDateTimeField(date, 'DAY_OF_MONTH', 31, tz)).toStrictEqual(new Date('2020-03-01T10:10:13'))
        expect(calculator.addDateTimeField(date, 'MONTH', 12, tz)).toStrictEqual(new Date('2021-01-30T10:10:13'))
        expect(calculator.addDateTimeField(date, 'DAY_OF_MONTH', 365, tz)).toStrictEqual(
            new Date('2021-01-29T10:10:13'),
        )

        expect(calculator.addDateTimeField(date, 'HOUR', 2, tz)).toStrictEqual(new Date('2020-01-30T12:10:13'))
        expect(calculator.addDateTimeField(date, 'HOUR', 26, tz)).toStrictEqual(new Date('2020-01-31T12:10:13'))
        expect(calculator.addDateTimeField(date, 'MINUTE', 61, tz)).toStrictEqual(new Date('2020-01-30T11:11:13'))
        expect(calculator.addDateTimeField(date, 'SECOND', 61, tz)).toStrictEqual(new Date('2020-01-30T10:11:14'))
    })
    it('should differentiate Date and DateTime', () => {
        const date = new Date('2020-01-30T10:10:13')
        expect(calculator.isDate(date)).toBeTruthy()
        expect(calculator.isDateTime(date)).toBeTruthy()
    })
    it('should throw if timezone not equal to system', () => {
        expect(() => calculator.today('Africa/Mbabane')).toThrow()
    })
})

File: kraken-expression-language/src/test/java/kraken/el/date/DefaultDateCalculatorTest.java
/*
 * Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kraken.el.date;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThrows;

import java.time.DateTimeException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeParseException;

import org.junit.Before;
import org.junit.Test;

import kraken.el.date.DateCalculator.DateField;
import kraken.el.date.DateCalculator.DateTimeField;
import kraken.utils.Dates;

/**
 * @author Mindaugas Ulevicius
 */
public class DefaultDateCalculatorTest {

    private final ZoneId tz = ZoneId.of("Europe/Berlin");

    private DefaultDateCalculator calculator;

    @Before
    public void setUp() throws Exception {
        this.calculator = new DefaultDateCalculator();
    }

    @Test
    public void shouldCreateDate() {
        assertThat(
            calculator.createDate(2020, 1, 23),
            equalTo(LocalDate.of(2020, 1, 23))
        );
        assertThrows(DateTimeException.class, () -> calculator.createDate(2020, 2, 30));
        assertThrows(DateTimeException.class, () -> calculator.createDate(2020, 0, 1));
        assertThrows(DateTimeException.class, () -> calculator.createDate(2020, 13, 1));
        assertThrows(DateTimeException.class, () -> calculator.createDate(2020, 1, 0));

        assertThat(
            calculator.createDate("2020-01-23"),
            equalTo(LocalDate.of(2020, 1, 23))
        );
        assertThrows(DateTimeException.class, () -> calculator.createDate("2020-02-30"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDate("2020-00-01"));
        assertThrows(DateTimeException.class, () -> calculator.createDate("2020-13-01"));
        assertThrows(DateTimeException.class, () -> calculator.createDate("2020-01-00"));

        assertThrows(DateTimeParseException.class, () -> calculator.createDate("2020-01-23T10:00:00"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDate("2020-01-23T10:00:00Z"));
    }

    @Test
    public void shouldCreateDateTime() {
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-23"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-23T10:00:00"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-02-30T10:00:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-00-01T10:00:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-13-01T10:00:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-00T10:00:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-23T25:00:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-23T10:60:00Z"));
        assertThrows(DateTimeParseException.class, () -> calculator.createDateTime("2020-01-23T10:00:60Z"));

        assertThat(
            calculator.createDateTime("2020-01-23T10:00:00", tz),
            equalTo(atLocalTz(ZonedDateTime.of(LocalDateTime.parse("2020-01-23T10:00:00"), tz)))
        );
        assertThat(
            calculator.createDateTime("2020-01-23T10:00:00Z"),
            equalTo(Dates.convertISOToLocalDateTime("2020-01-23T10:00:00Z"))
        );
    }

    @Test
    public void shouldReturnToday() {
        assertThat(calculator.today(ZoneId.systemDefault()), equalTo(LocalDate.now()));
        assertThat(calculator.today(tz), equalTo(atTz(LocalDateTime.now()).toLocalDate()));
    }

    @Test
    public void shouldConvertDateToDateTime() {
        var startOfDay = calculator.toDateTime(LocalDate.of(2020, 1, 23), ZoneId.systemDefault());
        assertThat(startOfDay, equalTo(LocalDateTime.of(2020, 1, 23, 0, 0, 0)));

        var startOfDayInVilnius = calculator.toDateTime(LocalDate.of(2020, 1, 23), tz);
        var expectedStartOfDayInVilnius = atLocalTz(LocalDate.of(2020, 1, 23).atStartOfDay(tz));

        assertThat(startOfDayInVilnius, equalTo(expectedStartOfDayInVilnius));
    }

    @Test
    public void shouldConvertDateTimeToDate() {
        var time = LocalDateTime.of(2020, 1, 23, 10, 0, 0);

        assertThat(calculator.toDate(time, ZoneId.systemDefault()), equalTo(LocalDate.of(2020, 1, 23)));
        assertThat(calculator.toDate(time, tz), equalTo(atTz(time).toLocalDate()));
    }

    @Test
    public void shouldGetDateField() {
        var date = LocalDate.of(2020, 1, 30);
        assertThat(calculator.getDateField(date, DateField.YEAR), equalTo(2020));
        assertThat(calculator.getDateField(date, DateField.MONTH), equalTo(1));
        assertThat(calculator.getDateField(date, DateField.DAY_OF_MONTH), equalTo(30));
    }

    @Test
    public void shouldSetDateField() {
        var date = LocalDate.of(2020, 1, 30);
        assertThat(calculator.withDateField(date, DateField.YEAR, 2021), equalTo(LocalDate.of(2021, 1, 30)));
        assertThat(calculator.withDateField(date, DateField.MONTH, 3), equalTo(LocalDate.of(2020, 3, 30)));
        assertThat(calculator.withDateField(date, DateField.MONTH, 2), equalTo(LocalDate.of(2020, 2, 29)));
        assertThat(calculator.withDateField(date, DateField.DAY_OF_MONTH, 10), equalTo(LocalDate.of(2020, 1, 10)));

        assertThrows(DateTimeException.class, () -> calculator.withDateField(date, DateField.MONTH, 13));
        assertThrows(DateTimeException.class, () -> calculator.withDateField(date, DateField.DAY_OF_MONTH, 32));
    }

    @Test
    public void shouldAddDateField() {
        var date = LocalDate.of(2020, 1, 30);
        assertThat(calculator.addDateField(date, DateField.YEAR, 1), equalTo(LocalDate.of(2021, 1, 30)));
        assertThat(calculator.addDateField(date, DateField.MONTH, 1), equalTo(LocalDate.of(2020, 2, 29)));
        assertThat(calculator.addDateField(date, DateField.MONTH, 12), equalTo(LocalDate.of(2021, 1, 30)));
        assertThat(calculator.addDateField(date, DateField.DAY_OF_MONTH, 31), equalTo(LocalDate.of(2020, 3, 1)));
        assertThat(calculator.addDateField(date, DateField.DAY_OF_MONTH, 365), equalTo(LocalDate.of(2021, 1, 29)));
    }

    @Test
    public void shouldDiffDateField() {
        assertThat(calculator.difference(d("2020-01-30"), d("2022-01-30"), DateField.YEAR), equalTo(2L));
        assertThat(calculator.difference(d("2020-01-30"), d("2022-01-29"), DateField.YEAR), equalTo(1L));
        assertThat(calculator.difference(d("2020-01-30"), d("2021-01-29"), DateField.YEAR), equalTo(0L));

        assertThat(calculator.difference(d("2020-01-29"), d("2020-02-29"), DateField.MONTH), equalTo(1L));
        assertThat(calculator.difference(d("2020-01-30"), d("2020-02-29"), DateField.MONTH), equalTo(0L));
        assertThat(calculator.difference(d("2020-01-30"), d("2020-03-01"), DateField.MONTH), equalTo(1L));
        assertThat(calculator.difference(d("2020-01-29"), d("2021-01-28"), DateField.MONTH), equalTo(11L));

        assertThat(calculator.difference(d("2020-01-29"), d("2020-02-29"), DateField.DAY_OF_MONTH), equalTo(31L));
    }

    @Test
    public void shouldGetDateTimeFieldInSystemTz() {
        var dateTime = LocalDateTime.of(2020, 1, 30, 10, 10, 13);
        var defaultTz = ZoneId.systemDefault();
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.YEAR, defaultTz), equalTo(2020));
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.MONTH, defaultTz), equalTo(1));
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.DAY_OF_MONTH, defaultTz), equalTo(30));
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.HOUR, defaultTz), equalTo(10));
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.MINUTE, defaultTz), equalTo(10));
        assertThat(calculator.getDateTimeField(dateTime, DateTimeField.SECOND, defaultTz), equalTo(13));
    }

    @Test
    public void shouldSetDateTimeFieldInSystemTz() {
        var dateTime = LocalDateTime.of(2020, 1, 30, 10, 10, 13);
        var defaultTz = ZoneId.systemDefault();

        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.YEAR, 2021, defaultTz),
            equalTo(LocalDateTime.of(2021, 1, 30, 10, 10, 13)));
        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.MONTH, 3, defaultTz),
            equalTo(LocalDateTime.of(2020, 3, 30, 10, 10, 13)));
        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.DAY_OF_MONTH, 10, defaultTz),
            equalTo(LocalDateTime.of(2020, 1, 10, 10, 10, 13)));
        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.HOUR, 23, defaultTz),
            equalTo(LocalDateTime.of(2020, 1, 30, 23, 10, 13)));
        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.MINUTE, 5, defaultTz),
            equalTo(LocalDateTime.of(2020, 1, 30, 10, 5, 13)));
        assertThat(
            calculator.withDateTimeField(dateTime, DateTimeField.SECOND, 0, defaultTz),
            equalTo(LocalDateTime.of(2020, 1, 30, 10, 10, 0)));

        assertThat(calculator.withDateTimeField(dateTime, DateTimeField.MONTH, 2, defaultTz),
            equalTo(LocalDateTime.of(2020, 2, 29, 10, 10, 13)));

        assertThrows(DateTimeException.class, () -> calculator.withDateTimeField(dateTime, DateTimeField.MONTH, 13, defaultTz));
        assertThrows(DateTimeException.class, () -> calculator.withDateTimeField(dateTime, DateTimeField.DAY_OF_MONTH, 32, defaultTz));
        assertThrows(DateTimeException.class, () -> calculator.withDateTimeField(dateTime, DateTimeField.HOUR, 24, defaultTz));
        assertThrows(DateTimeException.class, () -> calculator.withDateTimeField(dateTime, DateTimeField.MINUTE, 60, defaultTz));
        assertThrows(DateTimeException.class, () -> calculator.withDateTimeField(dateTime, DateTimeField.SECOND, 60, defaultTz));
    }

    @Test
    public void shouldAddDateTimeFieldInSystemTz() {
        var dateTime = LocalDateTime.of(2020, 1, 30, 10, 10, 13);
        var defaultTz = ZoneId.systemDefault();

        assertThat(
            calculator.addDateTimeField(dateTime, DateTimeField.YEAR, 1, defaultTz),
            equalTo(LocalDateTime.of(2021, 1, 30, 10, 10, 13)));
        assertThat(
            calculator.addDateTimeField(dateTime, DateTimeField.MONTH, 1, defaultTz),
            equalTo(LocalDateTime.of(2020, 2, 29, 10, 10, 13)));
        assertThat(
            calculator.addDateTimeField(dateTime, DateTimeField.DAY_OF_MONTH, 31, defaultTz),
            equalTo(LocalDateTime.of(2020, 3, 1, 10, 10, 13)));
        assertThat(
            calculator.addDateTimeField(dateTime, DateTimeField.MONTH, 12, defaultTz),
            equalTo(LocalDateTime.of(2021, 1, 30, 10, 10, 13)));
        assertThat(
            calculator.addDateTimeField(dateTime, DateTimeField.DAY_OF_MONTH, 365, defaultTz),
            equalTo(LocalDateTime.of(2021, 1, 29, 10, 10, 13)));
    }

    private LocalDate d(String iso) {
        return LocalDate.parse(iso);
    }

    private LocalDateTime atLocalTz(ZonedDateTime zonedDateTime) {
        return zonedDateTime.withZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();
    }

    private LocalDateTime atTz(LocalDateTime localDateTime) {
        return localDateTime.atZone(ZoneId.systemDefault()).withZoneSameInstant(tz).toLocalDateTime();
    }
}
